File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    static void findCombinations(int idx,String digits,HashMap<String,String> telbtn ,List<String> result,StringBuilder temp){
        if(idx == digits.length()){
            result.add(temp.toString());
            return;
        }

        String letters = telbtn.get(String.valueOf(digits.charAt(idx)));
        for(int i=0; i< letters.length(); i++){
            temp.append(letters.charAt(i));
            findCombinations(idx+1,digits,telbtn,result,temp);
            temp.deleteCharAt(temp.length() -1);
        }
    }
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0)    return new ArrayList<>();

        HashMap<String, String> telbtn = new HashMap<>();
        telbtn.put("2","abc");
        telbtn.put("3","def");
        telbtn.put("4","ghi");
        telbtn.put("5","jkl");
        telbtn.put("6","mno");
        telbtn.put("7","pqrs");
        telbtn.put("8","tuv");
        telbtn.put("9","wxyz");

        List<String> result = new ArrayList<>();
        StringBuilder temp = new StringBuilder();
        findCombinations(0,digits,telbtn,result,temp);
        return result;
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    const map = new Map();

    // Map a string key to an array of characters
    map.set("2", ["a","b","c"]);
    map.set("3", ["d","e","f"]);
    map.set("4", ["g","h","i"]);
    map.set("5", ["j","k","l"]);
    map.set("6", ["m","n","o"]);
    map.set("7", ["p","q","r","s"]);
    map.set("8", ["t","u","v"]);
    map.set("9", ["w","x","y","z"]);

    const result = [];
    generateletterCombinations(result,"",0,map,digits);
    return result;
};

function generateletterCombinations(result,current,index,map,digits){

    //Base case 
    if(index == digits.length){
        result.push(current);
        return;
    }

    const letters = map.get(digits[index]);
    for(let i = 0; i < letters.length; i++){
        generateletterCombinations(
            result,
            current + letters[i],
            index+1,
            map,
            digits
        );
    }
}